### PR TITLE
(PDB-4766) Query reports with type=any

### PR DESCRIPTION
**Pull request by oanatmaria**
A new way of querying for all reports was introduced. All reports are queried by including `type = "any"` in the query body.

### CODE REVIEW

**Review by IrimieBogdan — APPROVED** [review]

**Review by Filipovici-Andrei — APPROVED** [review]

File: test/puppetlabs/puppetdb/http/reports_test.clj [hunk 1]
```diff
@@ -179,6 +179,26 @@
...
 
+(deftest-http-app query-with-type-any
+  [[version endpoint] endpoints
+    method [:get :post]]
+
+  (let [basic (:basic reports)
+        _ (store-example-report! basic (now))
+        plan-report (-> (:basic2 reports)
+                        (assoc-in [:certname] "foo.bar")
+                        (assoc-in [:type] "plan"))
```
Comment (sebastian-miclea) [inline test/puppetlabs/puppetdb/http/reports_test.clj:190]: You can use `assoc` and remove the brackets for `:certname` and `:type`

**Comment by austb** [comment]
The lint job hit a transient error, I've put up a fix for that error here https://github.com/puppetlabs/puppetdb/pull/3488

**Review by austb — APPROVED** [review]
:+1: looks good to me. Going to do a little more manual testing and then I'll merge it.

**Review by austb — COMMENTED** [review]
I found an edge case when both of the special `"any"` filters are used without any other filters.
```
$ curl -X POST http://localhost:8080/pdb/query/v4 \
  -H 'Content-Type:application/json' \
  -d '{"query": "reports[] { type = \"any\" and node_state = \"any\" }"}'
ERROR: syntax error at or near ")"
```

The error appears to be caused by the `WHERE ()` at the end of the query
```
2021-05-14 12:32:23,873 INFO  [qtp1772130429-30] [p.p.query-eng] PDBQuery:27b99075-563e-4521-9e3b-92c18e2ec056:["from" "reports" ["extract" [] ["and" ["=" "type" "any"] ["=" "node_state" "any"]]]]
2021-05-14 12:32:23,886 INFO  [qtp1772130429-30] [p.p.query-eng] PDBQuery:27b99075-563e-4521-9e3b-92c18e2ec056:["WITH inactive_nodes AS (SELECT certname FROM certnames WHERE (deactivated IS NOT NULL AND deactivated > '2021-04-30T19:32:23.885Z') OR (expired IS NOT NULL and expired > '2021-04-30T19:32:23.885Z')), not_active_nodes AS (SELECT certname FROM certnames WHERE (deactivated IS NOT NULL OR expired IS NOT NULL)) SELECT reports.receive_time AS receive_time, reports.corrective_change AS corrective_change, reports.code_id AS code_id, reports.noop AS noop, reports.certname AS certname, reports.puppet_version AS puppet_version, encode(reports.hash::bytea, 'hex') AS hash, reports.cached_catalog_status AS cached_catalog_status, reports.transaction_uuid::text AS transaction_uuid, reports.configuration_version AS configuration_version, report_statuses.status AS status, reports.job_id AS job_id, producers.name AS producer, reports.catalog_uuid::text AS catalog_uuid, (SELECT row_to_json(t) FROM (SELECT coalesce(logs, CAST(logs_json AS jsonb)) AS data, format('/pdb/query/v4/reports/%s/logs', encode(hash::bytea, 'hex')) AS href) t) AS logs, reports.noop_pending AS noop_pending, reports.end_time AS end_time, reports.report_type AS type, environments.environment AS environment, reports.start_time AS start_time, (SELECT row_to_json(event_data) FROM (SELECT json_agg(row_to_json(t)) AS data, format('/pdb/query/v4/reports/%s/events', encode(hash::bytea, 'hex')) AS href FROM (SELECT re.status, re.timestamp, re.resource_type, re.resource_title, re.property, re.corrective_change, CAST(re.new_value AS jsonb), CAST(re.old_value AS jsonb), re.message, re.file, re.line, re.containment_path, re.containing_class, re.name FROM resource_events re WHERE reports.id = re.report_id) t) event_data) AS resource_events, (SELECT row_to_json(t) FROM (SELECT coalesce(metrics, CAST(metrics_json AS jsonb)) AS data, format('/pdb/query/v4/reports/%s/metrics', encode(hash::bytea, 'hex')) AS href) t) AS metrics, reports.producer_timestamp AS producer_timestamp, reports.report_format AS report_format FROM reports LEFT JOIN environments ON environments.id = reports.environment_id LEFT JOIN producers ON producers.id = reports.producer_id LEFT JOIN report_statuses ON reports.status_id = report_statuses.id WHERE ()"]
```

A similar query on the nodes endpoint
```
curl -X POST http://localhost:8080/pdb/query/v4 \
  -H 'Content-Type:application/json' \
  -d '{"query": "nodes[certname] { node_state = \"any\"}" }'
```

Results in the removal of the `WHERE` clause entirely.

```
2021-05-14 12:34:12,234 INFO  [qtp1772130429-36] [p.p.query-eng] PDBQuery:d008a171-0f0c-450f-822d-2cc632f2d822:["from" "nodes" ["extract" ["certname"] ["=" "node_state" "any"]]]
2021-05-14 12:34:12,241 INFO  [qtp1772130429-36] [p.p.query-eng] PDBQuery:d008a171-0f0c-450f-822d-2cc632f2d822:["WITH inactive_nodes AS (SELECT certname FROM certnames WHERE (deactivated IS NOT NULL AND deactivated > '2021-04-30T19:34:12.240Z') OR (expired IS NOT NULL and expired > '2021-04-30T19:34:12.240Z')), not_active_nodes AS (SELECT certname FROM certnames WHERE (deactivated IS NOT NULL OR expired IS NOT NULL)) SELECT certnames.certname AS certname FROM certnames LEFT JOIN catalogs ON catalogs.certname = certnames.certname LEFT JOIN factsets fs ON certnames.certname = fs.certname LEFT JOIN reports ON (certnames.certname = reports.certname AND certnames.latest_report_id = reports.id) LEFT JOIN environments catalog_environment ON catalog_environment.id = catalogs.environment_id LEFT JOIN report_statuses ON reports.status_id = report_statuses.id LEFT JOIN environments facts_environment ON facts_environment.id = fs.environment_id LEFT JOIN environments reports_environment ON reports_environment.id = reports.environment_id"]
```

**Review by austb — APPROVED** [review]

**Pull request merged**